integration-tests/test_products.py

```diff
@@ -0,0 +1,132 @@
+# Copyright (c) 2025 Red Hat, Inc.
+#
+# This software is licensed to you under the GNU General Public
+# License as published by the Free Software Foundation; either version
+# 2 of the License (GPLv2) or (at your option) any later version.
+# There is NO WARRANTY for this software, express or implied,
+# including the implied warranties of MERCHANTABILITY,
+# NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
+# have received a copy of GPLv2 along with this software; if not, see
+# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+#
+
+from constants import RHSM, RHSM_PRODUCTS, TEST_PRODUCT_CERT_PATHS
+from utils import loop_until, installed_products
+import json
+import logging
+import pytest
+import shutil
+from pathlib import Path
+from funcy import first
+
+logger = logging.getLogger(__name__)
+
+"""
+Integration test for DBus RHSM Products Object.
+
+See https://www.candlepinproject.org/docs/subscription-manager/dbus_objects.html#products
+for more details.
+
+Main usecases are presented in this file.
+"""
+
+# each call uses standard english locale
+locale = "en_US.UTF-8"
+
+
+def test_products_list_installed_products(any_candlepin, subman, test_config):
+    """
+    https://www.candlepinproject.org/docs/subscription-manager/dbus_objects.html
+    a method Products.ListInstalledProducts
+
+    - The method returns a list of installed products.
+    - it is a content of /etc/pki/product* directories.
+
+    This test verifies tha DBus method Products/ListIntalledProducts returns a list of installed products.
+    """
+    subman.register(
+        username=test_config.get("candlepin", "username"),
+        password=test_config.get("candlepin", "password"),
+        org=test_config.get("candlepin", "org"),
+    )
+    loop_until(lambda: subman.is_registered)
+
+    # testing certificates are in the same dir as test files (in a directory 'files/product')
+    src_cert_path = first(TEST_PRODUCT_CERT_PATHS)
+    dst_cert_path = Path("/etc/pki/product") / src_cert_path.name
+
+    # copy product certificate to ensure that at least one exists in the dir /etc/pki/product
+    shutil.copyfile(src_cert_path, dst_cert_path)
+
+    proxy = RHSM.get_proxy(RHSM_PRODUCTS)
+    response = json.loads(proxy.ListInstalledProducts("", {}, locale))
+    #
+    # [["Red Hat Enterprise Linux for x86_64 Beta", "486", "10.1 Beta", "x86_64", "unknown", [], "", ""]]
+    #
+
+    def values_with_keys(product_values: list[str]) -> dict[str, str]:
+        """convert into a dict using given keys"""
+        keys_in_response = ("Product Name", "Product ID", "Version", "Arch")
+        return dict(zip(keys_in_response, product_values))
+
+    products_in_response = [values_with_keys(product_values) for product_values in response]
+    products = installed_products(subman)
```

Comment:
No, you do not want to test it this way. When something is wrong with the shared code (subman and rhsm.service share the critical part of code), then you will get false positive results.

Please, compare result from D-Bus method with  the content of the directories (`/etc/pki/product` and `/etc/pki/product-default`).